### PR TITLE
Refactor self transcoder

### PR DESCRIPTION
**Pull request by basine**
`ZMSelfTranscoder` is now `ZMSelfStrategy` and works independent from the sync state machine.

Note: This is based on https://github.com/wireapp/wire-ios-sync-engine/pull/158 which needs to be merged first

### CODE REVIEW

File: Source/Synchronization/Sync States/ZMEventProcessingState.m [hunk 1]
```diff
@@ -24,16 +24,15 @@
...
 #import "ZMStateMachineDelegate.h"
-#import "ZMHotFix.h"
+#import <zmessaging/zmessaging-Swift.h>
 
 @interface ZMEventProcessingState ()
 
 @property (nonatomic) NSArray *syncObjects;
 @property (nonatomic) BOOL isSyncing; // Only used to send a notification to UI that syncing finished
-@property (nonatomic) ZMHotFix *hotFix;
+@property (nonatomic) SyncStatus *slowSynStatus;
```
Comment (daehn) [inline Source/Synchronization/Sync States/ZMEventProcessingState.m]: Typo `slowSynStatus` → `slowSyncStatus`

File: Source/Calling/ZMGSMCallHandler.m [hunk 2]
```diff
@@ -72,23 +73,23 @@ - (instancetype)initWithUIManagedObjectContext:(NSManagedObjectContext *)uiMOC
         self.callCenter.callEventHandler = self.callEventHandler;
         self.canUpdateCallState = NO;
         self.activeCallUIConversationObjectID = self.storedConversation.objectID;
-        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(didFinishSync:) name:ZMApplicationDidEnterEventProcessingStateNotificationName object:nil];
+        [ZMUserSession addInitalSyncCompletionObserver:self];
```
Comment (typfel) [inline Source/Calling/ZMGSMCallHandler.m]: typo: `addInitalSyncCompletionObserver ` >`addInitialSyncCompletionObserver `

File: Source/Calling/ZMGSMCallHandler.m [hunk 3]
```diff
@@ -72,23 +73,23 @@ - (instancetype)initWithUIManagedObjectContext:(NSManagedObjectContext *)uiMOC
...
     }
     return self;
 }
 
 - (void)dealloc
 {
-    [[NSNotificationCenter defaultCenter] removeObserver:self];
+    [ZMUserSession removeInitalSyncCompletionObserver:self];
```
Comment (typfel) [inline Source/Calling/ZMGSMCallHandler.m]: typo: `removeInitalSyncCompletionObserver` > `removeInitialSyncCompletionObserver `

File: Source/Synchronization/Sync States/ZMSyncStateMachine.m [hunk 4]
```diff
@@ -179,15 +157,10 @@ - (void)prepareForSuspendedState;
...
-- (void)startQuickSync
-{
-    [self goToState:self.updateEventsCatchUpPhaseOneState];
-}
-
-- (void)startSlowSync
-{
-    [self goToState:self.downloadLastUpdateEventIDState];
-}
+//- (void)startQuickSync
```
Comment (typfel) [inline Source/Synchronization/Sync States/ZMSyncStateMachine.m]: Delete?

File: Source/Synchronization/Transcoders/ZMConversationTranscoder.m [hunk 5]
```diff
@@ -160,6 +168,33 @@ - (BOOL)isSlowSyncDone
...
+    return SyncPhaseFetchingConversations;
+}
+
+- (ZMTransportRequest *)nextRequest
+{
+    if (!self.clientRegistrationDelegate.clientIsReadyForRequests) {
+        return nil;
+    }
+    
+    SyncStatus *status = self.syncStatus;
```
Comment (typfel) [inline Source/Synchronization/Transcoders/ZMConversationTranscoder.m]: Is this temporary code that will be re-factored in another PR? It seems to be repeated in a a few transcoders.

File: Source/Synchronization/Transcoders/ZMSelfStrategy.m [hunk 6]
```diff
@@ -105,16 +105,28 @@ - (void)tearDown
...
         [self.timedDownstreamSync readyForNextRequestIfNotBusy];
-        return @[self.timedDownstreamSync];
+        return [self.timedDownstreamSync nextRequest];
     }
-    if (! self.isSlowSyncDone) {
-        return @[self.downstreamSelfUserSync];
+    
+    if (clientStatus.currentPhase == ZMClientRegistrationPhaseWaitingForSelfUser) {
+        ZMUser *selfUser = [ZMUser selfUserInContext:self.managedObjectContext];
+        if (!self.isSelfUserComplete && !selfUser.needsToBeUpdatedFromBackend) {
```
Comment (typfel) [inline Source/Synchronization/Transcoders/ZMSelfStrategy.m]: Is this logic necessary? Why would you ever be in the `ZMClientRegistrationPhaseWaitingForSelfUser ` phase if the `selfUser.needsToBeUpdatedFromBackend` is not `YES`?

File: Source/Calling/ZMGSMCallHandler.m (hunk 2 above)
Comment (basine) [inline Source/Calling/ZMGSMCallHandler.m]: That's a typo in the original function that I did not add.

File: Source/Synchronization/Sync States/ZMSyncStateMachine.m (hunk 4 above)
Comment (basine) [inline Source/Synchronization/Sync States/ZMSyncStateMachine.m]: Yeah, I am not done with all the refactoring. There might be ToDos that I still need to remove / resolve.

File: Source/Synchronization/Transcoders/ZMConversationTranscoder.m (hunk 5 above)
Comment (basine) [inline Source/Synchronization/Transcoders/ZMConversationTranscoder.m]: Yes

**Comment by basine** [comment]
@typfel exactly - this is WIP since I full refactor in one PR would probably be a 5000 line change and ain't nobody wants to review this

File: Source/Calling/ZMGSMCallHandler.m (hunk 2 above)
Comment (typfel) [inline Source/Calling/ZMGSMCallHandler.m]: Ah I see. still nice to fix typos as we discover them, broken window ...

File: Source/Synchronization/Transcoders/ZMSelfStrategy.m (hunk 6 above)
Comment (basine) [inline Source/Synchronization/Transcoders/ZMSelfStrategy.m]: Yes, previously we would only set the flag needsToBeUpdatedFromBackend once we left `ZMUnauthenticatedState` after finishing registration. It's rather the self.isSelfUserComplete check that is duplicating what `ZMClientRegistrationPhaseWaitingForSelfUser` is reflecting.

**Pull request closed without merging**